// sober/dist/segmented-button.js
import { builder, html } from './core/element.js';
import './ripple.js';
const style = /*css*/ `
:host{
  display: inline-flex;
  align-items: center;
  vertical-align: middle;
  border: solid 1px var(--s-color-outline, #777680);
  border-radius: 20px;
  height: 40px;
  overflow: hidden;
}
`;
const name = 's-segmented-button';
const props = {};
export class SegmentedButton extends builder({
    name, style, props,
    setup() {
        let options = [];
        let selectedIndex = -1;
        let changed = false;
        const slotChange = (_, el) => {
            options = el.assignedElements().filter((item) => item instanceof SegmentedButtonItem);
            selectedIndex = -1;
            let target = null;
            for (const item of options) {
                if (item.selected)
                    target = item;
            }
            if (target)
                update(target);
        };
        const update = (target) => {
            if (options.length === 0 || !target.selected)
                return (selectedIndex = -1);
            let old = null;
            for (const item of options) {
                if (item === target)
                    continue;
                if (item.selected) {
                    old = item;
                    item.removeAttribute('selected');
                }
            }
            selectedIndex = options.indexOf(target);
            if (changed) {
                this.dispatchEvent(new Event('change'));
                changed = false;
            }
        };
        this.addEventListener('segmented-button-item:update', (event) => {
            event.stopPropagation();
            update(event.target);
        });
        this.addEventListener('segmented-button-item:change', (event) => {
            event.stopPropagation();
            changed = true;
        });
        return {
            expose: {
                get options() {
                    return options;
                },
                get selectedIndex() {
                    return selectedIndex;
                }
            },
            render: () => html `
        <slot @slotchange="${slotChange}"></slot>
      `
        };
    }
}) {
}
const itemStyle = /*css*/ `
:host{
  flex-basis: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--s-color-on-surface, #1c1b1f);
  height: 100%;
  min-width: 48px;
  padding: 0 16px;
  text-transform: capitalize;
  font-weight: 500;
  font-size: .875rem;
  position: relative;
  cursor: pointer;
  box-sizing: border-box;
  border-left: solid 1px var(--s-color-outline, #777680);
}
:host(:first-child){
  border-left-color: transparent;
  margin-left: -1px;
}
:host([selected=true]){
  background: var(--s-color-secondary-container, #e2e0f9);
  color: var(--s-color-on-secondary-container, #191a2c);
}
:host([disabled=true]){
  pointer-events: none;
  color: color-mix(in srgb, var(--s-color-on-surface, #1c1b1f) 38%, transparent);
}
::slotted(s-icon){
  width: 18px;
  height: 18px;
  flex-shrink: 0;
}
::slotted(s-icon[slot=start]){
  margin-right: 4px;
}
::slotted(s-icon[slot=end]){
  margin-right: 4px;
}
`;
const itemName = 's-segmented-button-item';
const itemProps = {
    selected: false,
    disabled: false,
    selectable: true
};
export class SegmentedButtonItem extends builder({
    name: itemName,
    style: itemStyle,
    props: itemProps,
    propSyncs: ['selected', 'disabled'],
    setup() {
        this.addEventListener('click', () => {
            if (this.selectable) {
                if (this.selected)
                    return;
                if (this.parentNode instanceof SegmentedButton) {
                    this.dispatchEvent(new Event('navigation-item:change', { bubbles: true }));
                }
                this.selected = true;
            }
        });
        return {
            watches: {
                selected: () => {
                    if (!(this.parentNode instanceof SegmentedButton))
                        return;
                    this.dispatchEvent(new Event('segmented-button-item:update', { bubbles: true }));
                }
            },
            render: () => html `
        <slot name="start"></slot>
        <slot></slot>
        <slot name="end"></slot>
        <s-ripple class="ripple" attached="true"></s-ripple>
      `
        };
    }
}) {
}
SegmentedButton.define();
SegmentedButtonItem.define();
